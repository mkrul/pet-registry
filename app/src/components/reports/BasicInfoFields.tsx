import React from "react";
import { TextField, SelectChangeEvent } from "@mui/material";
import { IReportForm } from "../../types/Report";

const commonInputStyles = {
  backgroundColor: "white",
  "& .MuiOutlinedInput-root": {
    backgroundColor: "white"
  }
};

interface BasicInfoFieldsProps {
  formData: IReportForm;
  onInputChange?: (
    e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement> | SelectChangeEvent
  ) => void;
  readOnly?: boolean;
}

export const BasicInfoFields: React.FC<BasicInfoFieldsProps> = ({
  formData,
  onInputChange,
  readOnly
}) => {
  return (
    <>
      <TextField
        label="Title"
        name="title"
        value={formData.title || ""}
        onChange={onInputChange}
        variant="outlined"
        fullWidth
        disabled={readOnly}
        sx={commonInputStyles}
      />

      <TextField
        label="Description"
        name="description"
        value={formData.description || ""}
        onChange={onInputChange}
        multiline
        rows={4}
        variant="outlined"
        fullWidth
        disabled={readOnly}
        sx={commonInputStyles}
      />

      <TextField
        label="Pet's Name (leave blank if not known)"
        name="name"
        value={formData.name || ""}
        onChange={onInputChange}
        variant="outlined"
        fullWidth
        disabled={readOnly}
        sx={commonInputStyles}
      />
    </>
  );
};
