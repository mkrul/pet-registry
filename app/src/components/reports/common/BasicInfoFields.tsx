import React from "react";
import { TextField } from "@mui/material";
import { BasicInfoFieldsProps } from "../../../types/Report";
import { commonInputStyles } from "../../../styles/commonStyles";
import Tip from "../../common/Tip";
import { FormFieldError } from "../../common/FormFieldError";

export const BasicInfoFields: React.FC<BasicInfoFieldsProps> = ({
  formData,
  onInputChange,
  readOnly,
  error,
  descriptionError
}) => {
  return (
    <div className="space-y-6">
      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Title:</label>
        <TextField
          name="title"
          value={formData.title || ""}
          onChange={onInputChange}
          variant="outlined"
          fullWidth
          required
          disabled={readOnly}
          sx={commonInputStyles}
          error={!!error}
        />
        <FormFieldError error={error} />
      </div>

      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Pet's Name:</label>{" "}
        <span className="text-sm text-gray-500"> (Leave blank if not known)</span>
        <TextField
          name="name"
          value={formData.name || ""}
          onChange={onInputChange}
          variant="outlined"
          fullWidth
          disabled={readOnly}
          sx={commonInputStyles}
        />
      </div>

      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Description:</label>
        <Tip>Describe the animal's appearance, behavior, and other relevant details.</Tip>
        <TextField
          name="description"
          value={formData.description || ""}
          onChange={onInputChange}
          multiline
          rows={4}
          variant="outlined"
          fullWidth
          required
          disabled={readOnly}
          sx={commonInputStyles}
          error={!!descriptionError}
        />
        <FormFieldError error={descriptionError} />
      </div>
    </div>
  );
};
