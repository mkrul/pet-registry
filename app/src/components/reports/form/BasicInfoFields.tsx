import React from "react";
import { TextField } from "@mui/material";
import { BasicInfoFieldsProps } from "../../../types/Report";
import { commonInputStyles } from "../../../styles/commonStyles";

export const BasicInfoFields: React.FC<BasicInfoFieldsProps> = ({
  formData,
  onInputChange,
  readOnly
}) => {
  return (
    <>
      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Title:</label>
        <TextField
          name="title"
          value={formData.title || ""}
          onChange={onInputChange}
          variant="outlined"
          fullWidth
          required
          disabled={readOnly}
          sx={commonInputStyles}
        />
      </div>

      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Description:</label>
        <TextField
          name="description"
          value={formData.description || ""}
          onChange={onInputChange}
          multiline
          rows={4}
          variant="outlined"
          fullWidth
          required
          disabled={readOnly}
          sx={commonInputStyles}
        />
      </div>

      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900 mb-2">Pet's Name (if known):</label>
        <TextField
          name="name"
          value={formData.name || ""}
          onChange={onInputChange}
          variant="outlined"
          fullWidth
          disabled={readOnly}
          sx={commonInputStyles}
        />
      </div>
    </>
  );
};
