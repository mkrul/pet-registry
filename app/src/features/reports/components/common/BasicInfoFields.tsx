import React from "react";
import { TextField } from "@mui/material";
import { BasicInfoFieldsProps } from "../../types/Report";
import { commonInputStyles } from "../../../../shared/commonStyles";
import Tip from "../../../../shared/components/common/Tip";
import { FormFieldError } from "../../../../shared/components/common/FormFieldError";

export const BasicInfoFields:  = ({
  formData,
  onInputChange,
  readOnly,
  error,
  descriptionError
}) => {
  return (
    <div className="space-y-6">
      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900">Pet Name:</label>
        <span className="text-sm text-gray-500"> (Leave blank if not known)</span>
        <TextField
          name="name"
          value={formData.name}
          onChange={onInputChange}
          variant="outlined"
          fullWidth
          disabled={readOnly}
          sx={commonInputStyles}
        />
      </div>

      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900">Title:</label>
        <TextField
          name="title"
          value={formData.title}
          onChange={onInputChange}
          variant="outlined"
          fullWidth
          multiline
          rows={1}
          required
          disabled={readOnly}
          error={!!error}
          sx={commonInputStyles}
        />
        <FormFieldError error={error} />
      </div>

      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900">Description:</label>
        <TextField
          name="description"
          value={formData.description}
          onChange={onInputChange}
          variant="outlined"
          fullWidth
          multiline
          rows={4}
          required
          disabled={readOnly}
          error={!!descriptionError}
          sx={commonInputStyles}
        />
        <FormFieldError error={descriptionError} />
      </div>
    </div>
  );
};
