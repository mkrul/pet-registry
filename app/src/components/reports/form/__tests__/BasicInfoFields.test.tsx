import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom/vitest";
import { BasicInfoFields } from "../BasicInfoFields";
import type { BasicInfoFieldsProps } from "../../../../types/Report";

const mockFormData = {
  title: "Test Title",
  description: "Test Description",
  name: "Test Pet Name" as string | null,
  species: "Dog",
  breed1: "Labrador",
  breed2: null,
  color1: "Black",
  color2: null,
  color3: null,
  gender: "Male",
  microchipId: "123456",
  latitude: 0,
  longitude: 0,
  area: "Test Area",
  state: "Test State",
  country: "Test Country",
  image: {
    id: "1",
    url: "/test-image.jpg",
    thumbnailUrl: "/test-image-thumb.jpg",
    variantUrl: "/test-image.jpg",
    filename: "test-image.jpg",
    publicId: "test-image"
  }
};

describe("BasicInfoFields", () => {
  const mockOnInputChange = vi.fn();

  const mockProps: Required<BasicInfoFieldsProps> = {
    formData: mockFormData,
    onInputChange: mockOnInputChange,
    readOnly: false
  };

  beforeEach(() => {
    mockOnInputChange.mockClear();
  });

  describe("Rendering", () => {
    it("should match snapshot", () => {
      const { container } = render(<BasicInfoFields {...mockProps} />);
      expect(container).toMatchSnapshot();
    });

    it.each([
      ["title", "Test Title"],
      ["description", "Test Description"],
      ["pet's name", "Test Pet Name"]
    ])("should render %s field with value '%s'", (label, value) => {
      render(<BasicInfoFields {...mockProps} />);
      const input = screen.getByLabelText(new RegExp(label, "i"));
      expect(input).toBeDefined();
      expect(input).toHaveValue(value);
    });

    it.each([
      ["title", ""],
      ["description", ""],
      ["pet's name", ""]
    ])("should handle empty %s value", (label, value) => {
      render(
        <BasicInfoFields
          {...mockProps}
          formData={{
            ...mockFormData,
            [label === "pet's name" ? "name" : label]: value
          }}
        />
      );
      expect(screen.getByLabelText(new RegExp(label, "i"))).toHaveValue(value);
    });
  });

  describe("User Interactions", () => {
    it.each([
      ["title", "New Title"],
      ["description", "New Description"],
      ["pet's name", "New Name"]
    ])("should handle %s input change", async (label, newValue) => {
      const user = userEvent.setup();
      render(<BasicInfoFields {...mockProps} />);

      const input = screen.getByLabelText(new RegExp(label, "i"));

      // Directly simulate change event instead of clear + type
      await act(async () => {
        const fieldName = label === "pet's name" ? "name" : label;
        const changeEvent = {
          target: {
            name: fieldName,
            value: newValue
          }
        };
        if (mockProps.onInputChange) {
          mockProps.onInputChange(changeEvent as React.ChangeEvent<HTMLInputElement>);
        }
      });

      expect(mockOnInputChange).toHaveBeenCalledWith(
        expect.objectContaining({
          target: expect.objectContaining({
            name: label === "pet's name" ? "name" : label,
            value: newValue
          })
        })
      );
    });

    it("should handle rapid input changes", async () => {
      const user = userEvent.setup();
      render(<BasicInfoFields {...mockProps} />);

      const titleInput = screen.getByLabelText(/title/i);
      const newValue = "Test Title - Updated";

      await act(async () => {
        const changeEvent = {
          target: {
            name: "title",
            value: newValue
          }
        };
        if (mockProps.onInputChange) {
          mockProps.onInputChange(changeEvent as React.ChangeEvent<HTMLInputElement>);
        }
      });

      expect(mockOnInputChange).toHaveBeenLastCalledWith(
        expect.objectContaining({
          target: expect.objectContaining({
            name: "title",
            value: newValue
          })
        })
      );
    });
  });

  describe("ReadOnly Mode", () => {
    it.each([["title"], ["description"], ["pet's name"]])(
      "should disable %s input in readonly mode",
      label => {
        render(<BasicInfoFields {...mockProps} readOnly={true} />);
        expect(screen.getByLabelText(new RegExp(label, "i"))).toBeDisabled();
      }
    );

    it("should prevent input changes when readonly", async () => {
      const user = userEvent.setup();
      render(<BasicInfoFields {...mockProps} readOnly={true} />);

      const titleInput = screen.getByLabelText(/title/i);
      await act(async () => {
        await user.type(titleInput, "Should not change");
      });

      expect(mockOnInputChange).not.toHaveBeenCalled();
    });
  });

  describe("Error Handling", () => {
    it("should handle null values gracefully", () => {
      render(
        <BasicInfoFields
          {...mockProps}
          formData={{
            ...mockFormData,
            title: "",
            description: "",
            name: null
          }}
        />
      );

      expect(screen.getByLabelText(/title/i)).toHaveValue("");
      expect(screen.getByLabelText(/description/i)).toHaveValue("");
      expect(screen.getByLabelText(/pet's name/i)).toHaveValue("");
    });
  });
});
