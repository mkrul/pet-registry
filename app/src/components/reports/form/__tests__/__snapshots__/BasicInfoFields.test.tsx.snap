// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BasicInfoFields > Rendering > should match snapshot 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-tuk39d-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-113d811-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for=":r0:"
      id=":r0:-label"
    >
      Title
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1blp12k-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r0:"
        name="title"
        type="text"
        value="Test Title"
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-w1u3ce"
        >
          <span>
            Title
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-tuk39d-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-113d811-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for=":r1:"
      id=":r1:-label"
    >
      Description
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-zsp0bh-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <textarea
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-w4nesw-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r1:"
        name="description"
        rows="4"
        style="height: 0px; overflow: hidden;"
      >
        Test Description
      </textarea>
      <textarea
        aria-hidden="true"
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-w4nesw-MuiInputBase-input-MuiOutlinedInput-input"
        readonly=""
        style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
        tabindex="-1"
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-w1u3ce"
        >
          <span>
            Description
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-tuk39d-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-113d811-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for=":r2:"
      id=":r2:-label"
    >
      Pet's Name (leave blank if not known)
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1blp12k-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r2:"
        name="name"
        type="text"
        value="Test Pet Name"
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-w1u3ce"
        >
          <span>
            Pet's Name (leave blank if not known)
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
